File: wpimath/src/main/native/cpp/geometry/struct/Translation2dStruct.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/geometry/struct/Translation2dStruct.hpp"

namespace {
constexpr size_t X_OFF = 0;
constexpr size_t Y_OFF = X_OFF + 8;
}  // namespace

using StructType = wpi::util::Struct<wpi::math::Translation2d>;

wpi::math::Translation2d StructType::Unpack(std::span<const uint8_t> data) {
  return wpi::math::Translation2d{
      wpi::units::meters<>{wpi::util::UnpackStruct<double, X_OFF>(data)},
      wpi::units::meters<>{wpi::util::UnpackStruct<double, Y_OFF>(data)},
  };
}

void StructType::Pack(std::span<uint8_t> data,
                      const wpi::math::Translation2d& value) {
  wpi::util::PackStruct<X_OFF>(data, value.X().value());
  wpi::util::PackStruct<Y_OFF>(data, value.Y().value());
}

File: wpimath/src/main/java/org/wpilib/math/geometry/struct/Translation2dStruct.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.geometry.struct;

import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;

public class Translation2dStruct implements Struct<Translation2d> {
  @Override
  public Class<Translation2d> getTypeClass() {
    return Translation2d.class;
  }

  @Override
  public String getTypeName() {
    return "Translation2d";
  }

  @Override
  public int getSize() {
    return kSizeDouble * 2;
  }

  @Override
  public String getSchema() {
    return "double x;double y";
  }

  @Override
  public Translation2d unpack(ByteBuffer bb) {
    double x = bb.getDouble();
    double y = bb.getDouble();
    return new Translation2d(x, y);
  }

  @Override
  public void pack(ByteBuffer bb, Translation2d value) {
    bb.putDouble(value.getX());
    bb.putDouble(value.getY());
  }

  @Override
  public boolean isImmutable() {
    return true;
  }
}
